### ISSUE TITLE
Label Reviewer Not loading 

### ISSUE THREAD

**Issue by AHarouni** [Project-MONAI/MONAILabel#1300]
**Describe the bug**
Label Reviewer plugin Doesn't work anymore:
- on unstabel nightly, plug in is not installed with monai label
- on stabel slicer 5.2.1, it hangs on when clicking load 
![image](https://user-images.githubusercontent.com/61221417/218536257-31513919-30a3-48a6-a51e-db7ffe591980.png)
- on prevouse slicer 5.1.0 I can load the list of images but can't load any of the images themselves 

**Server logs**
No logs on the server after it starts on hots
```
[INFO] (uvicorn.error:206) - Uvicorn running on http://0.0.0.0:8000/ (Press CTRL+C to quit)
```
on slicer logs 
```
Python 3.9.10 (main, Aug  9 2022, 23:27:28) [MSC v.1930 64 bit (AMD64)] on win32
>>> 2023-02-13 09:57:37.708310: Search input field is empty

2023-02-13 09:56:17.050577: Cache is cleaned
2023-02-13 09:56:17.145325: Successfully connected to server (server url: 'http://clarasa-station:8000').
2023-02-13 09:56:17.220126: Selected filters: segmented= False | isNotSegmented= False | isApproved= False | isFlagged= False
2023-02-13 09:56:17.226123: Successfully loaded Image data [total = 41, category = 'All']
2023-02-13 09:56:17.349776: Request Data  image_name='spleen_24.nii.gz', node_name='spleen_24.nii.gz', image_id='spleen_24', checksum=''
2023-02-13 09:56:17.356757: REST: request dicom image 'http://clarasa-station:8000/datastore/image?image=spleen_24'
<b>Download failed (attempt 1 of 3)...</b>
<b>Download failed (attempt 2 of 3)...</b>
<b>Download failed (attempt 3 of 3)...</b>
Traceback (most recent call last):
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 152, in index_changed
    self.loadImageData()
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 412, in loadImageData
    self.loadNextImage(self.currentImageData)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 979, in loadNextImage
    self.logic.loadDicomAndSegmentation(imageData)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 1134, in loadDicomAndSegmentation
    self.requestDicomImage(image_id, image_name, node_name, checksum)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 1174, in requestDicomImage
    _volumeNode = sampleDataLogic.downloadFromURL(
IndexError: list index out of range
2023-02-13 09:56:17.429565: Selected reviewer: ''
2023-02-13 09:56:46.059733: Cache is cleaned
2023-02-13 09:56:46.119630: Successfully connected to server (server url: 'http://clarasa-station:8000').
2023-02-13 09:56:46.189250: Selected filters: segmented= False | isNotSegmented= False | isApproved= False | isFlagged= False
2023-02-13 09:56:46.195234: Successfully loaded Image data [total = 41, category = 'All']
2023-02-13 09:56:46.281624: Request Data  image_name='spleen_24.nii.gz', node_name='spleen_24.nii.gz', image_id='spleen_24', checksum=''
2023-02-13 09:56:46.288652: REST: request dicom image 'http://clarasa-station:8000/datastore/image?image=spleen_24'
<b>Download failed (attempt 1 of 3)...</b>
<b>Download failed (attempt 2 of 3)...</b>
<b>Download failed (attempt 3 of 3)...</b>
Traceback (most recent call last):
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 152, in index_changed
    self.loadImageData()
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 412, in loadImageData
    self.loadNextImage(self.currentImageData)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 979, in loadNextImage
    self.logic.loadDicomAndSegmentation(imageData)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 1134, in loadDicomAndSegmentation
    self.requestDicomImage(image_id, image_name, node_name, checksum)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 1174, in requestDicomImage
    _volumeNode = sampleDataLogic.downloadFromURL(
IndexError: list index out of range
2023-02-13 09:56:53.565766: Selected filters: segmented= False | isNotSegmented= False | isApproved= False | isFlagged= False
2023-02-13 09:56:53.573112: Successfully loaded Image data [total = 41, category = 'All']
2023-02-13 09:56:53.664913: Request Data  image_name='spleen_24.nii.gz', node_name='spleen_24.nii.gz', image_id='spleen_24', checksum=''
2023-02-13 09:56:53.670895: REST: request dicom image 'http://clarasa-station:8000/datastore/image?image=spleen_24'
<b>Download failed (attempt 1 of 3)...</b>
<b>Download failed (attempt 2 of 3)...</b>
<b>Download failed (attempt 3 of 3)...</b>
Traceback (most recent call last):
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 412, in loadImageData
    self.loadNextImage(self.currentImageData)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 979, in loadNextImage
    self.logic.loadDicomAndSegmentation(imageData)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 1134, in loadDicomAndSegmentation
    self.requestDicomImage(image_id, image_name, node_name, checksum)
  File "C:/Users/aharouni/AppData/Local/NA-MIC/Slicer 5.1.0-2022-08-08/NA-MIC/Extensions-31106/MONAILabel/lib/Slicer-5.1/qt-scripted-modules/MONAILabelReviewer.py", line 1174, in requestDicomImage
    _volumeNode = sampleDataLogic.downloadFromURL(
IndexError: list index out of range
```

**To Reproduce**
1. start radiology sample app with spleen sample dataset
2. open reviewlabel plugin 
3. click load

**Comment by jamesobutler** [comment]
> on unstabel nightly, plug in is not installed with monai label

@AHarouni Can you describe more about what is going on here^. Does this happen on another machine? I installed latest Slicer Preview 5.3.0-2023-02-10, installed the MONAILabel extension from the ExtensionManager, restarted Slicer, and observed that the MONAILabel Reviewer module was loaded.
![image](https://user-images.githubusercontent.com/15837524/218626993-d4a7f748-2957-4d3b-b35d-a7ab55ff3f4b.png)

**Comment by AHarouni** [comment]
Thanks for the quick response. 
I usually have 2 or 3 slicer installed, I couldn't find the monai review in the preview installation I installed early Jan. So I assumed it is only in the stable build. 
I also with slicer on linux I only see the monai label without the reviewer 
![image](https://user-images.githubusercontent.com/61221417/218637124-7481b756-65e8-4832-abc5-93e5ca466725.png)

However, you proved me wrong at least for windows so I guess I was unlikey and downloaded on a bad day where the plugin was not built or something. 
Anyway I am downloading a preview of 2023-2-11 for windows today and the monai review shows up with monai label. Thanks

Unfortunately, I get error connecting to the monai label. 
![image](https://user-images.githubusercontent.com/61221417/218636807-6057cc10-f2d3-4a03-84fb-1afdc01d9761.png)

Although with monai label I can load and get next images 
![image](https://user-images.githubusercontent.com/61221417/218636870-44365750-ed87-44a4-8438-3e07c5ab0886.png)

**Comment by jlvahldiek** [comment]
I can confirm that the Reviewer is crashing and not working any more, this was also recently discussed here #1290. I am working on a quick fix for the moment. But we should rethink a complete re-write of the Reviewer which was initially committed just as a prototype. As far as I understood, @chezhia might be willing to contribute.

**Comment by DaoMaoDuc** [comment]
Thanks a lot for raising the issue.

@jlvahldiek is correct. Since the frame work in MONAILabel is evolving continuously, the MONAILabelReviewer has to be adjusted as well according to the code changes in MONAILabel.

I figure out the issue:
In the log above we can see the information "checksum" is missing.
` Request Data  image_name='spleen_24.nii.gz', node_name='spleen_24.nii.gz', image_id='spleen_24', checksum=''`

That is why I checked out the code of monailabel. Referring to the current implementation of "LocalDatastoreModel" which you can find in ".../MONAILabel/monailabel/datastore/local.py" you can see that the code line 
` image_info["checksum"] = file_checksum(dest)` is commented out (the same occurred in the method "save_label" as well). Consequently, the information of checksum will not be persisted. However, that information should be passed to the request in order to load the label/segmentation properly.

`    def add_image(self, image_id: str, image_filename: str, image_info: Dict[str, Any]) -> str:
        id, image_ext = self._to_id(os.path.basename(image_filename))
        if not image_id:
            image_id = id

        logger.info(f"Adding Image: {image_id} => {image_filename}")
        name = self._filename(image_id, image_ext)
        dest = os.path.realpath(os.path.join(self._datastore.image_path(), name))

        with FileLock(self._lock_file):
            logger.debug("Acquired the lock!")
            shutil.copy(image_filename, dest)

            image_info = image_info if image_info else {}
            image_info["ts"] = int(time.time())
            **# image_info["checksum"] = file_checksum(dest)**
            image_info["name"] = name

            self._datastore.objects[image_id] = ImageLabelModel(image=DataModel(info=image_info, ext=image_ext))
            self._update_datastore_file(lock=False)
        logger.debug("Released the lock!")
        return image_id`

Probably we should ask the authors why they commented out that line.

**Comment by jlvahldiek** [comment]
I just started PR #1307. Two issues are causing the extension to crash. I think that I fixed one of it (removed checksum-related code).

The other issue is still unsolved: The extension crashes during `slicer.mrmlScene.Clear()`. As far as I understood, this is related to the integration of the SegmentEditorWidget. I pasted the related stack trace from the crash report below. Maybe @lassoan or @jamesobutler can see the problem.

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libMRMLCore.dylib             	       0x1106063e1 std::__1::__tree_node_base<void*>*& std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<vtkMRMLHierarchyNode*, std::__1::allocator<vtkMRMLHierarchyNode*> > >, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<vtkMRMLHierarchyNode*, std::__1::allocator<vtkMRMLHierarchyNode*> > >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<vtkMRMLHierarchyNode*, std::__1::allocator<vtkMRMLHierarchyNode*> > > > >::__find_equal<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(std::__1::__tree_end_node<std::__1::__tree_node_base<void*>*>*&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 17
1   libMRMLCore.dylib             	       0x1106d0928 vtkMRMLNode::GetNthNodeReference(char const*, int) + 200
2   libvtkSlicerSegmentationsModuleMRML.dylib	       0x1525c7792 vtkMRMLSegmentEditorNode::GetSegmentationNode() + 18
3   libqSlicerSegmentationsModuleWidgets.dylib	       0x152fcbb97 qMRMLSegmentEditorWidget::onMRMLSceneEndBatchProcessEvent() + 39
4   QtCore                        	       0x10f440075 0x10f214000 + 2277493
5   libCTKVisualizationVTKCore.0.1.dylib	       0x10d09ed43 ctkVTKConnection::emitExecute(vtkObject*, void*, unsigned long, void*) + 99
6   libCTKVisualizationVTKCore.0.1.dylib	       0x10d097db8 ctkVTKConnectionPrivate::execute(vtkObject*, unsigned long, void*, void*) + 168
7   libvtkCommon-9.1.1.dylib      	       0x12847a1b1 vtkCallbackCommand::Execute(vtkObject*, unsigned long, void*) + 33
8   libMRMLCore.dylib             	       0x11060006b vtkEventBroker::InvokeObservation(vtkObservation*, unsigned long, void*) + 171
9   libMRMLCore.dylib             	       0x1105ffc10 vtkEventBroker::ProcessEvent(vtkObservation*, vtkObject*, unsigned long, void*) + 144
10  libvtkCommon-9.1.1.dylib      	       0x12847a1b1 vtkCallbackCommand::Execute(vtkObject*, unsigned long, void*) + 33
11  libvtkCommon-9.1.1.dylib      	       0x12858e2c0 vtkSubjectHelper::InvokeEvent(unsigned long, void*, vtkObject*) + 1200
12  libMRMLCore.dylib             	       0x1106f66f4 vtkMRMLScene::EndState(unsigned long) + 740
13  libqSlicerBaseQTApp.dylib     	       0x10c55b15d qSlicerMainWindow::on_FileCloseSceneAction_triggered() + 109
14  libqSlicerBaseQTApp.dylib     	       0x10c56571c qSlicerMainWindow::qt_metacall(QMetaObject::Call, int, void**) + 76
15  libqSlicerApp.dylib           	       0x10c46be78 qSlicerAppMainWindow::qt_metacall(QMetaObject::Call, int, void**) + 24
16  QtCore                        	       0x10f43ff0e 0x10f214000 + 2277134
17  QtWidgets                     	       0x10d9f53f6 QAction::activate(QAction::ActionEvent) + 310
18  QtCore                        	       0x10f437b3f QObject::event(QEvent*) + 943
19  QtWidgets                     	       0x10d9fea1a QApplicationPrivate::notify_helper(QObject*, QEvent*) + 266
20  QtWidgets                     	       0x10d9ffe41 QApplication::notify(QObject*, QEvent*) + 497
21  libqSlicerBaseQTGUI.dylib     	       0x10ca03910 qSlicerApplication::notify(QObject*, QEvent*) + 16
22  QtCore                        	       0x10f40c9d4 QCoreApplication::notifyInternal2(QObject*, QEvent*) + 212
23  QtCore                        	       0x10f40dd19 QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) + 809
24  libqcocoa.dylib               	       0x1231d8259 0x12319e000 + 238169
25  libqcocoa.dylib               	       0x1231d89c8 0x12319e000 + 240072
26  CoreFoundation                	    0x7ff812ef617b __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
27  CoreFoundation                	    0x7ff812ef60e3 __CFRunLoopDoSource0 + 180
28  CoreFoundation                	    0x7ff812ef5ebf __CFRunLoopDoSources0 + 340
29  CoreFoundation                	    0x7ff812ef4878 __CFRunLoopRun + 892
30  CoreFoundation                	    0x7ff812ef3e3c CFRunLoopRunSpecific + 562
31  HIToolbox                     	    0x7ff81bba35e6 RunCurrentEventLoopInMode + 292
32  HIToolbox                     	    0x7ff81bba3213 ReceiveNextEventCommon + 283
33  HIToolbox                     	    0x7ff81bba30e5 _BlockUntilNextEventMatchingListInModeWithFilter + 70
34  AppKit                        	    0x7ff81592dfad _DPSNextEvent + 927
35  AppKit                        	    0x7ff81592c66a -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1394
36  AppKit                        	    0x7ff81591ed19 -[NSApplication run] + 586
37  libqcocoa.dylib               	       0x1231d762f 0x12319e000 + 235055
38  QtCore                        	       0x10f408a6f QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 431
39  QtCore                        	       0x10f40cfe2 QCoreApplication::exec() + 130
40  libqSlicerBaseQTCore.dylib    	       0x10d324f49 qSlicerCoreApplication::exec() + 9
41  Slicer                        	       0x10c23bae3 main + 483
42  dyld                          	       0x11236d52e start + 462
```

**Comment by jlvahldiek** [comment]
When setting up the MRMLScene for the SegmentEditorWidget, the extension crashes on `slicer.mrmlScene.Clear()` with the above stack trace.

The MRMLscene gets set up here:
https://github.com/Project-MONAI/MONAILabel/blob/cbb859cddc7adff7307bc882a638df9fbbee5ed1/plugins/slicer/MONAILabelReviewer/MONAILabelReviewer.py#L1385

When the above line is commented out, calling `slicer.mrmlScene.Clear()` does not crash Slicer, but the SegmentEditorWidget does not work anymore.

Any ideas? Thanks.

**Comment by DaoMaoDuc** [comment]
I guess you cannot just omit the checksum, since there is an important checksum verification subroutine within the slicer implementation.  I guess that happens here `extractAlgoAndDigest(checksum)` which you can find in util.py in the slicer repo.

According to its implementation, `extractAlgoAndDigest` returns None if the checksum is None.

**Comment by lassoan** [comment]
> The other issue is still unsolved: The extension crashes during slicer.mrmlScene.Clear().

I can definitely help with this. Could you provide a minimum set of steps that reproduces this issue?

**Comment by jlvahldiek** [comment]
> I can definitely help with this. Could you provide a minimum set of steps that reproduces this issue?

Steps to reproduce:
1. Install MONAILabel extension from Extension Manager in Slicer 5.2.1 (crash also occurs in 5.3.0)
2. Open MONAILabelReviewer module in Slicer
3. Click  `Menu` -> `File` -> `Close Scene` or execute `slicer.mrmlscene.Clear()` in Python Interactor

**Comment by lassoan** [comment]
I've fixed the crash on scene close.

It was due to that the parameter node was not set in the Segment Editor widget. I've submitted a [pull request to fix this](https://github.com/Project-MONAI/MONAILabel/pull/1312).

Of course, incorrect use of the widget must not ever cause application crash. I've [fixed this](https://github.com/Slicer/Slicer/commit/9506e42834364ad528f5e26c0150ca25e4c44999) in the Slicer Preview Release today.
